File: solutions/0501. Find Mode in Binary Search Tree/0501.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def findMode(self, root: Optional[TreeNode]) -> List[int]:
    count = {}
        
    def inorder(node):
      if node:
        inorder(node.left)
        count[node.val] = count.get(node.val, 0) + 1
        inorder(node.right)

    inorder(root)
    max_count = max(count.values())
    
    return [val for val, freq in count.items() if freq == max_count]

File: solutions/0501. Find Mode in Binary Search Tree/0501.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  vector<int> findMode(TreeNode* root) {
    unordered_map<int, int> count;

    inorder(root, count);

    int max_count = 0;
    for (const auto& [val, freq] : count)
      max_count = max(max_count, freq);

    vector<int> ans;
    for (const auto& [val, freq] : count)
      if (freq == max_count)
        ans.push_back(val);

    return ans;
  }

 private:
  void inorder(TreeNode* node, unordered_map<int, int>& count) {
    if (node == nullptr)
      return;

    inorder(node->left, count);
    ++count[node->val];
    inorder(node->right, count);
  }
};